zstd/seqdec_amd64.go

```diff
@@ -9,6 +9,147 @@ import (
 	"github.com/klauspost/compress/internal/cpuinfo"
 )
 
+type decodeSyncAsmContext struct {
+	llTable     []decSymbol
+	mlTable     []decSymbol
+	ofTable     []decSymbol
+	llState     uint64
+	mlState     uint64
+	ofState     uint64
+	iteration   int
+	litRemain   int
+	out         []byte
+	outPosition int
+	literals    []byte
+	litPosition int
+	history     []byte
+	windowSize  int
+	retry       bool // set by the caller when `out` got resized after reporting errorOutOfCapacity
+	ll          int  // set on error (not for all errors, please refer to _generate/gen.go)
+	ml          int  // set on error (not for all errors, please refer to _generate/gen.go)
+	mo          int  // set on error (not for all errors, please refer to _generate/gen.go)
+}
+
+// sequenceDecs_decodeSync_amd64 implements the main loop of sequenceDecs.decodeSync in x86 asm.
+//
+// Please refer to seqdec_generic.go for the reference implementation.
+//go:noescape
+func sequenceDecs_decodeSync_amd64(s *sequenceDecs, br *bitReader, ctx *decodeSyncAsmContext) int
+
+// sequenceDecs_decodeSync_bmi2 implements the main loop of sequenceDecs.decodeSync in x86 asm with BMI2 extensions.
+//go:noescape
+func sequenceDecs_decodeSync_bmi2(s *sequenceDecs, br *bitReader, ctx *decodeSyncAsmContext) int
+
+// decode sequences from the stream with the provided history but without a dictionary.
+func (s *sequenceDecs) decodeSyncSimple(hist []byte) (bool, error) {
+	if len(s.dict) > 0 {
+		return false, nil
+	}
```

Comment:
This fixes the hanging:
```suggestion
	if len(s.dict) > 0 || cap(s.out) < maxCompressedBlockSize+maxCompressedBlockSizeAlloc {
		return false, nil
	}
```